File: vandalizer-next/backend/app/routers/workflows.py
"""Workflow API routes."""

import base64
import csv
import io
import json
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse

from app.dependencies import get_current_user
from app.models.user import User
from app.schemas.workflows import (
    AddStepRequest,
    AddTaskRequest,
    CreateWorkflowRequest,
    ReorderStepsRequest,
    RunWorkflowRequest,
    TestStepRequest,
    UpdateStepRequest,
    UpdateTaskRequest,
    UpdateWorkflowRequest,
    ValidateWorkflowRequest,
    ValidateWorkflowResponse,
    WorkflowResponse,
    WorkflowStatusResponse,
)
from app.services import workflow_service as svc

router = APIRouter()


# ---------------------------------------------------------------------------
# Workflow CRUD
# ---------------------------------------------------------------------------

@router.post("", response_model=WorkflowResponse)
async def create_workflow(req: CreateWorkflowRequest, user: User = Depends(get_current_user)):
    wf = await svc.create_workflow(req.name, user.user_id, req.space, req.description)
    return WorkflowResponse(
        id=str(wf.id), name=wf.name, description=wf.description,
        user_id=wf.user_id, space=wf.space, num_executions=wf.num_executions,
    )


@router.get("", response_model=list[WorkflowResponse])
async def list_workflows(space: str | None = None, user: User = Depends(get_current_user)):
    workflows = await svc.list_workflows(space=space)
    return [
        WorkflowResponse(
            id=str(wf.id), name=wf.name, description=wf.description,
            user_id=wf.user_id, space=wf.space, num_executions=wf.num_executions,
        )
        for wf in workflows
    ]


@router.get("/status", response_model=WorkflowStatusResponse)
async def get_workflow_status(session_id: str, user: User = Depends(get_current_user)):
    status = await svc.get_workflow_status(session_id)
    if not status:
        raise HTTPException(status_code=404, detail="Workflow result not found")
    return WorkflowStatusResponse(**status)


@router.get("/steps/test/{task_id}")
async def poll_step_test(task_id: str, user: User = Depends(get_current_user)):
    return svc.get_test_status(task_id)


@router.get("/download")
async def download_results(
    session_id: str,
    format: str = "json",
    user: User = Depends(get_current_user),
):
    """Download workflow results in specified format."""
    status = await svc.get_workflow_status(session_id)
    if not status:
        raise HTTPException(status_code=404, detail="Workflow result not found")

    final_output = status.get("final_output", {})
    output_data = final_output.get("output", "") if isinstance(final_output, dict) else final_output

    # Check for file_download type (e.g., from DataExport or DocumentRenderer)
    if isinstance(output_data, dict) and output_data.get("type") == "file_download":
        file_bytes = base64.b64decode(output_data["data_b64"])
        media_type_map = {"pdf": "application/pdf", "csv": "text/csv", "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "json": "application/json", "zip": "application/zip"}
        media_type = media_type_map.get(output_data.get("file_type", ""), "application/octet-stream")
        return StreamingResponse(
            io.BytesIO(file_bytes),
            media_type=media_type,
            headers={"Content-Disposition": f'attachment; filename="{output_data.get("filename", "output")}"'},
        )

    if format == "csv":
        buf = io.StringIO()
        writer = csv.writer(buf)
        if isinstance(output_data, list):
            if output_data and isinstance(output_data[0], dict):
                headers = list(output_data[0].keys())
                writer.writerow(headers)
                for row in output_data:
                    writer.writerow([row.get(h, "") for h in headers])
            else:
                writer.writerow(["Value"])
                for item in output_data:
                    writer.writerow([str(item)])
        elif isinstance(output_data, dict):
            writer.writerow(["Key", "Value"])
            for k, v in output_data.items():
                writer.writerow([k, str(v)])
        else:
            writer.writerow(["Output"])
            writer.writerow([str(output_data)])
        return StreamingResponse(
            io.BytesIO(buf.getvalue().encode()),
            media_type="text/csv",
            headers={"Content-Disposition": 'attachment; filename="results.csv"'},
        )

    # Default: JSON
    json_bytes = json.dumps(output_data, indent=2, default=str).encode()
    return StreamingResponse(
        io.BytesIO(json_bytes),
        media_type="application/json",
        headers={"Content-Disposition": 'attachment; filename="results.json"'},
    )


@router.get("/{workflow_id}", response_model=WorkflowResponse)
async def get_workflow(workflow_id: str, user: User = Depends(get_current_user)):
    wf = await svc.get_workflow(workflow_id)
    if not wf:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return WorkflowResponse(**wf)


@router.patch("/{workflow_id}", response_model=WorkflowResponse)
async def update_workflow(workflow_id: str, req: UpdateWorkflowRequest, user: User = Depends(get_current_user)):
    wf = await svc.update_workflow(workflow_id, name=req.name, description=req.description)
    if not wf:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return WorkflowResponse(
        id=str(wf.id), name=wf.name, description=wf.description,
        user_id=wf.user_id, space=wf.space, num_executions=wf.num_executions,
    )


@router.delete("/{workflow_id}")
async def delete_workflow(workflow_id: str, user: User = Depends(get_current_user)):
    ok = await svc.delete_workflow(workflow_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return {"ok": True}


@router.post("/{workflow_id}/duplicate", response_model=WorkflowResponse)
async def duplicate_workflow(workflow_id: str, user: User = Depends(get_current_user)):
    wf = await svc.duplicate_workflow(workflow_id, user.user_id)
    if not wf:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return WorkflowResponse(**wf)


# ---------------------------------------------------------------------------
# Steps
# ---------------------------------------------------------------------------

@router.post("/{workflow_id}/steps")
async def add_step(workflow_id: str, req: AddStepRequest, user: User = Depends(get_current_user)):
    step = await svc.add_step(workflow_id, req.name, req.data, req.is_output)
    if not step:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return step


@router.patch("/steps/{step_id}")
async def update_step(step_id: str, req: UpdateStepRequest, user: User = Depends(get_current_user)):
    step = await svc.update_step(step_id, name=req.name, data=req.data, is_output=req.is_output)
    if not step:
        raise HTTPException(status_code=404, detail="Step not found")
    return step


@router.delete("/steps/{step_id}")
async def delete_step(step_id: str, user: User = Depends(get_current_user)):
    ok = await svc.delete_step(step_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Step not found")
    return {"ok": True}


# ---------------------------------------------------------------------------
# Tasks
# ---------------------------------------------------------------------------

@router.post("/steps/{step_id}/tasks")
async def add_task(step_id: str, req: AddTaskRequest, user: User = Depends(get_current_user)):
    task = await svc.add_task(step_id, req.name, req.data)
    if not task:
        raise HTTPException(status_code=404, detail="Step not found")
    return task


@router.patch("/tasks/{task_id}")
async def update_task(task_id: str, req: UpdateTaskRequest, user: User = Depends(get_current_user)):
    task = await svc.update_task(task_id, name=req.name, data=req.data)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.delete("/tasks/{task_id}")
async def delete_task(task_id: str, user: User = Depends(get_current_user)):
    ok = await svc.delete_task(task_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Task not found")
    return {"ok": True}


# ---------------------------------------------------------------------------
# Execution
# ---------------------------------------------------------------------------

@router.post("/{workflow_id}/run")
async def run_workflow(workflow_id: str, req: RunWorkflowRequest, user: User = Depends(get_current_user)):
    try:
        session_id = await svc.run_workflow(
            workflow_id, req.document_uuids, user.user_id, req.model
        )
        return {"session_id": session_id}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/steps/test")
async def test_step(req: TestStepRequest, user: User = Depends(get_current_user)):
    task_id = await svc.test_step(
        req.task_name, req.task_data, req.document_uuids, user.user_id, req.model
    )
    return {"task_id": task_id}


@router.post("/{workflow_id}/reorder-steps")
async def reorder_steps(workflow_id: str, req: ReorderStepsRequest, user: User = Depends(get_current_user)):
    ok = await svc.reorder_steps(workflow_id, req.step_ids)
    if not ok:
        raise HTTPException(status_code=400, detail="Invalid step IDs or workflow not found")
    return {"ok": True}


@router.post("/{workflow_id}/validate", response_model=ValidateWorkflowResponse)
async def validate_workflow(workflow_id: str, req: ValidateWorkflowRequest, user: User = Depends(get_current_user)):
    try:
        result = await svc.validate_workflow(workflow_id, req.eval_plan)
        return ValidateWorkflowResponse(**result)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: vandalizer-next/frontend/src/api/workflows.ts
import { apiFetch } from './client'
import type { Workflow, WorkflowStatus } from '../types/workflow'

// Workflow CRUD

export function createWorkflow(data: { name: string; space?: string; description?: string }) {
  return apiFetch<Workflow>('/api/workflows', { method: 'POST', body: JSON.stringify(data) })
}

export function listWorkflows(space?: string) {
  const params = space ? `?space=${encodeURIComponent(space)}` : ''
  return apiFetch<Workflow[]>(`/api/workflows${params}`)
}

export function getWorkflow(id: string) {
  return apiFetch<Workflow>(`/api/workflows/${id}`)
}

export function updateWorkflow(id: string, data: { name?: string; description?: string }) {
  return apiFetch<Workflow>(`/api/workflows/${id}`, { method: 'PATCH', body: JSON.stringify(data) })
}

export function deleteWorkflow(id: string) {
  return apiFetch<{ ok: boolean }>(`/api/workflows/${id}`, { method: 'DELETE' })
}

export function duplicateWorkflow(id: string) {
  return apiFetch<Workflow>(`/api/workflows/${id}/duplicate`, { method: 'POST' })
}

// Steps

export function addStep(workflowId: string, data: { name: string; data?: Record<string, unknown>; is_output?: boolean }) {
  return apiFetch(`/api/workflows/${workflowId}/steps`, { method: 'POST', body: JSON.stringify(data) })
}

export function updateStep(stepId: string, data: { name?: string; data?: Record<string, unknown>; is_output?: boolean }) {
  return apiFetch(`/api/workflows/steps/${stepId}`, { method: 'PATCH', body: JSON.stringify(data) })
}

export function deleteStep(stepId: string) {
  return apiFetch<{ ok: boolean }>(`/api/workflows/steps/${stepId}`, { method: 'DELETE' })
}

// Tasks

export function addTask(stepId: string, data: { name: string; data?: Record<string, unknown> }) {
  return apiFetch(`/api/workflows/steps/${stepId}/tasks`, { method: 'POST', body: JSON.stringify(data) })
}

export function updateTask(taskId: string, data: { name?: string; data?: Record<string, unknown> }) {
  return apiFetch<{ id: string; name: string; data: Record<string, unknown> }>(`/api/workflows/tasks/${taskId}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  })
}

export function deleteTask(taskId: string) {
  return apiFetch<{ ok: boolean }>(`/api/workflows/tasks/${taskId}`, { method: 'DELETE' })
}

// Execution

export function runWorkflow(workflowId: string, data: { document_uuids: string[]; model?: string }) {
  return apiFetch<{ session_id: string }>(`/api/workflows/${workflowId}/run`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function getWorkflowStatus(sessionId: string) {
  return apiFetch<WorkflowStatus>(`/api/workflows/status?session_id=${encodeURIComponent(sessionId)}`)
}

export function streamWorkflowStatus(
  sessionId: string,
  onStatus: (status: WorkflowStatus) => void,
  onError?: (err: unknown) => void,
): () => void {
  let aborted = false
  const controller = new AbortController()

  const url = `/api/workflows/status/stream?session_id=${encodeURIComponent(sessionId)}`

  ;(async () => {
    try {
      const res = await fetch(url, {
        credentials: 'include',
        signal: controller.signal,
      })
      if (!res.ok || !res.body) {
        onError?.(new Error('Failed to connect to workflow status stream'))
        return
      }

      const reader = res.body.getReader()
      const decoder = new TextDecoder()
      let buffer = ''

      while (!aborted) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        for (const line of lines) {
          if (!line.startsWith('data: ')) continue
          try {
            const data = JSON.parse(line.slice(6))
            if (data.error === 'not_found') {
              onError?.(new Error('Workflow result not found'))
              return
            }
            onStatus(data as WorkflowStatus)
            if (data.status === 'completed' || data.status === 'error' || data.status === 'failed') {
              return
            }
          } catch {
            // skip malformed events
          }
        }
      }
    } catch (err) {
      if (!aborted) onError?.(err)
    }
  })()

  // Return cleanup function
  return () => {
    aborted = true
    controller.abort()
  }
}

export function testStep(data: { task_name: string; task_data: Record<string, unknown>; document_uuids: string[]; model?: string }) {
  return apiFetch<{ task_id: string }>('/api/workflows/steps/test', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function getTestStepStatus(taskId: string) {
  return apiFetch<{ status: string; result?: unknown }>(`/api/workflows/steps/test/${taskId}`)
}

export function downloadResults(sessionId: string, format: string = 'json') {
  return `/api/workflows/download?session_id=${encodeURIComponent(sessionId)}&format=${format}`
}

// Step reordering

export function reorderSteps(workflowId: string, stepIds: string[]) {
  return apiFetch<{ ok: boolean }>(`/api/workflows/${workflowId}/reorder-steps`, {
    method: 'POST',
    body: JSON.stringify({ step_ids: stepIds }),
  })
}

// Validation

export interface ValidationCheck {
  name: string
  status: 'PASS' | 'FAIL' | 'WARN' | 'SKIP'
  detail: string | null
}

export interface ValidationResult {
  grade: string
  summary: string
  checks: ValidationCheck[]
}

export function validateWorkflow(workflowId: string, evalPlan?: string, textInput?: string) {
  return apiFetch<ValidationResult>(`/api/workflows/${workflowId}/validate`, {
    method: 'POST',
    body: JSON.stringify({ eval_plan: evalPlan, text_input: textInput }),
  })
}
